pint_pandas/pint_array.py

```diff
@@ -386,7 +404,17 @@ def units(self):
 
     @property
     def quantity(self):
-        return self._Q(self.data, self._dtype.units)
+        data = self.data
+        if data.dtype in dtypeunmap:
+           try:
+               data = data.astype(dtypeunmap[data.dtype])
+           except:
+               # We might get here for integer arrays with <NA> values
+               # In that case, the returned quantity will have dtype=O, which is less useful.
+               pass
+        if hasattr(data, "to_numpy"):
```

Comment:
I am unsure on this - it feels inconsistent to sometimes return a quantity with a numpy array magnitude, sometimes a pandas array magnitude.

Is a quantity with a pandas array magnitude preferred over a quantity with a numpy object array magnitude?
Would it be better to always return pandas array magnitude?

Should add some tests for these different cases (once we've decided the prefered approach) so we don't change this in the future.

Reply:
data.astype() returns a numpy array, not a pd array, so you can't do to_numpy on it. This method should always return a Quantity of a numpy array, however the dtype of the array can be a pd dtype in cases where your data contains <NA>, but is a np dtype when it does not. This gives you the most functionality you can get in both cases.